### ISSUE TITLE
checking "Users successfully connect" against the actual user

### ISSUE THREAD

**Issue by giuliohome** [ploeh/UserManagement#5]
Since I like symmetry, when I read "Users successfully connect" I assume they connect "to each other"...
So I would setup a stricter test<sup>1</sup>

```haskell
        return $
          isOK actual &&
          (any (elem otherUser . connectedUsers) (Map.lookup (userId user) db)) &&
          (any (elem user . connectedUsers) (Map.lookup (userId otherUser) db)) &&
          ( elem user $ connectedUsers $ extractUser actual)
```
which fails until one fixes

```haskell
  connect <- runExceptT $ do
      user <- ExceptT $ return userRes
      otherUser <- ExceptT $ return otherUserRes
      lift $ updateUser $ addConnection user otherUser
      let otherUserUpdated = addConnection otherUser user
      lift $ updateUser otherUserUpdated
      return otherUserUpdated
```

(1): having defined:
```haskell
extractUser :: HttpResponse User -> User 
extractUser (OK user) = user
```

**Comment by giuliohome** [comment]
And by the way, since I would expect to represent that user1 is connected to user2 if user2 is connected to user1, I'd define the  connectedUsers as [Integer] not as [User], to solve the cyclic graph representation problem!

**Comment by ploeh** [comment]
Yes, one could change the problem to have that definition, but it wasn't what I had in mind. I had in mind something like Twitter. E.g. I follow Martin Fowler, but he doesn't follow me.

**Comment by giuliohome** [comment]
I guess that Twitter API returns the updated user and not the other one :-p , but I'm too lazy to check it<sup>(1)</sup>, even if I've used also that specific api in my old Android app. I thought I should rewrite it in Kotlin or making it a web app and so... why not in Haskell (yesod?) ?
Thanks for sharing. And sorry for being such a pest ;))  

(1) I've checked and they return the followed user in case of success: no wonder their apis are ugly and not an example of functional programming!)) But in any case they avoid a cyclic graph representation and their GET friendships/show includes only the screen names of the friends (not the friends of the friends and so on...)

**Comment by giuliohome** [comment]
TL;DR: Testing for mutual friendship is counter-intuitive 

The following naive test fails
```haskell
      testProperty "Mutual friendship" $ \
        user otherUser -> runStateTest $ do
        put $ Map.fromList [toDBEntry user, toDBEntry otherUser]
        actual1 <- post lookupUser updateUser (show $ userId user) (show $ userId otherUser)
        actual2 <- post lookupUser updateUser (show $ userId otherUser) (show $ userId user) 
        db <- get
        return $
          isOK actual1 && isOK actual2 &&
          (any (elem otherUser . connectedUsers) (Map.lookup (userId user) db)) &&
          (any (elem user . connectedUsers) (Map.lookup (userId otherUser) db))
```
due to the connected users representation and one has to resort to getting the user again from the db, in a quite cumbersome way since the returned actual user is not the updated one. 
```haskell
      testProperty "Mutual friendship" $ \
        user otherUser -> runStateTest $ do
        put $ Map.fromList [toDBEntry user, toDBEntry otherUser]
        actual1 <- post lookupUser updateUser (show $ userId user) (show $ userId otherUser)
        db <- get
        let updatedUser = fromJust $ Map.lookup (userId user) db
        actual2 <- post lookupUser updateUser (show $ userId otherUser) (show $ userId user) -- updatedUser)
        db <- get
        return $
          isOK actual1 && isOK actual2 &&
          (any (elem otherUser . connectedUsers) (Map.lookup (userId user) db)) &&
          (any (elem updatedUser . connectedUsers) (Map.lookup (userId otherUser) db))
```
But IMHO the best way to fix is to test only on the userId, excluding the connectedUsers part (which I also think should be reduced to `[Int]`)
So, in the connection test, `user` should become `(userId user)` and `connectedUsers` should become `(map userId . connectedUsers)` resulting in 
```haskell
      testProperty "Mutual friendship" $ \
        user otherUser -> runStateTest $ do
        put $ Map.fromList [toDBEntry user, toDBEntry otherUser]
        actual1 <- post lookupUser updateUser (show $ userId user) (show $ userId otherUser)
        actual2 <- post lookupUser updateUser (show $ userId otherUser) (show $ userId user) 
        db <- get
        return $
          isOK actual1 && isOK actual2 &&
          (any (elem (userId otherUser) . (map userId . connectedUsers)) (Map.lookup (userId user) db)) &&
          (any (elem (userId user) . (map userId . connectedUsers)) (Map.lookup (userId otherUser) db))
```
or, similarly, for the "Users successfully connect" Property:
```haskell
      testProperty "Users successfully connect" $ \
        user otherUser -> runStateTest $ do
        put $ Map.fromList [toDBEntry user, toDBEntry otherUser]
        actual <- post lookupUser updateUser (show $ userId user) (show $ userId otherUser)
        db <- get
        return $
          isOK actual &&
          (any (elem (userId otherUser) . (map userId . connectedUsers)) (Map.lookup (userId user) db))
```
